applications/serial_lte_modem/src/http_c/slm_at_httpc.c

```diff
@@ -420,23 +238,116 @@ static int payload_cb(int sock, struct http_request *req, void *user_data)
 
 static int do_http_connect(void)
 {
-	/* Connect to server if it is not connected yet. */
-	if (httpc.fd == INVALID_SOCKET) {
-		httpc.fd = server_connect(httpc.host, httpc.sec_tag);
-		if (httpc.fd == INVALID_SOCKET) {
-			LOG_ERR("server_connect fail.");
-			sprintf(rsp_buf, "\r\n#XHTTPCCON: 0\r\n");
-			rsp_send(rsp_buf, strlen(rsp_buf));
-		} else {
-			sprintf(rsp_buf, "\r\n#XHTTPCCON: 1\r\n");
-			rsp_send(rsp_buf, strlen(rsp_buf));
-		}
-	} else {
+	int ret;
+
+	if (httpc.fd != INVALID_SOCKET) {
 		LOG_ERR("Already connected to server.");
 		return -EINVAL;
 	}
 
-	return httpc.fd;
+	/* Open socket */
+	if (httpc.sec_tag == INVALID_SEC_TAG) {
+		ret = socket(httpc.family, SOCK_STREAM, IPPROTO_TCP);
+	} else {
+		ret = socket(httpc.family, SOCK_STREAM, IPPROTO_TLS_1_2);
+
+	}
+	if (ret < 0) {
+		LOG_ERR("socket() failed: %d", -errno);
+		return ret;
+	}
+	httpc.fd = ret;
+
+	/* Set socket options */
+	const uint32_t timeout_ms = HTTPC_REQ_TO_S * MSEC_PER_SEC;
+	struct timeval timeo = {
+		.tv_sec = (timeout_ms / 1000),
+		.tv_usec = (timeout_ms % 1000) * 1000,
+	};
+
+	LOG_DBG("Configuring socket timeout (%ld s)", timeo.tv_sec);
+	ret = setsockopt(httpc.fd, SOL_SOCKET, SO_SNDTIMEO, &timeo, sizeof(timeo));
+	if (ret) {
+		LOG_ERR("setsockopt(SO_SNDTIMEO) error: %d", -errno);
+		ret = -errno;
+		goto exit_cli;
+	}
+	ret = setsockopt(httpc.fd, SOL_SOCKET, SO_RCVTIMEO, &timeo, sizeof(timeo));
+	if (ret) {
+		LOG_ERR("setsockopt(SO_SNDTIMEO) error: %d", -errno);
+		ret = -errno;
+		goto exit_cli;
+	}
+	if (httpc.sec_tag != INVALID_SEC_TAG) {
+		sec_tag_t sec_tag_list[] = { httpc.sec_tag };
+		int peer_verify = TLS_PEER_VERIFY_REQUIRED;
+
+		ret = setsockopt(httpc.fd, SOL_TLS, TLS_SEC_TAG_LIST, sec_tag_list,
+				 sizeof(sec_tag_t));
+		if (ret) {
+			LOG_ERR("setsockopt(TLS_SEC_TAG_LIST) error: %d", -errno);
+			ret = -errno;
+			goto exit_cli;
+		}
+		ret = setsockopt(httpc.fd, SOL_TLS, TLS_PEER_VERIFY, &peer_verify,
+				 sizeof(peer_verify));
+		if (ret) {
+			LOG_ERR("setsockopt(TLS_PEER_VERIFY) error: %d", -errno);
+			ret = -errno;
+			goto exit_cli;
+		}
+		ret = setsockopt(httpc.fd, SOL_TLS, TLS_HOSTNAME, httpc.host, sizeof(httpc.host));
```

Comment:
This should rather be `strlen(httpc.host) + 1` instead of `sizeof(httpc.host)` - you should provide the actual string length (with NULL terminator) instead of the actual buffer length.

Reply:
Yes this is a bug.
Thanks for pointing it out.